ecocrumb.1
ECOCRUMB(1)                 User Commands                 ECOCRUMB(1)

NAME
       ecocrumb - Process target directories, convert PDFs to plaintext,
       generate topics_features.json and parsed_sections.json

SYNOPSIS
       ecocrumb [--force] [--batch] [--train] <target_directory> [num_topics]

DESCRIPTION
       The ecocrumb script is a utility for processing target directories,
       converting PDFs to plaintext files, and generating topics_features.json
       and parsed_sections.json files. It provides options to force
       regeneration of files, batch process multiple folders, and run the
       generate_topics.sh script.

OPTIONS
       --force
              Forces regeneration of the topics_features.json and
              parsed_sections.json files, even if they already exist.

       --batch
              Processes each folder within the target_directory separately.

       --train
              Enables the generation of topics_features.json using the
              generate_topics.sh script.

ARGUMENTS
       <target_directory>
              The directory or path to the target directory to be processed.

       [num_topics]
              The number of topics to generate. Default value: 15.

       Note:   The square brackets [] indicate optional arguments.

EXAMPLES
       1. Process a single target directory:
          $ ecocrumb target_dir

       2. Process a target directory with a specific number of topics:
          $ ecocrumb target_dir 10

       3. Force regeneration and process a target directory:
          $ ecocrumb --force target_dir

       4. Batch process multiple folders within a target directory:
          $ ecocrumb --batch target_dir

       5. Enable training mode and process a target directory:
          $ ecocrumb --train target_dir

NOTES
       - The script requires the esgetlpipeline scripts to be present in the
         same directory as the ecocrumb script.

       - The output directories 'topics_features' and 'parsed_sections' will
         be created within the target_directory.

       - The 'plaintext' directory will be created within the target_directory
         to store converted plaintext files.

       - The 'results/' and 'esg_LanguageModel/src/' directories will be
         created within the directory where the ecocrumb script is located,
         and the parsed_sections files will be copied to those directories.

SEE ALSO
       generate_topics.sh(1), parse_by_topic.sh(1)

AUTHOR
       Written by Your Name.

COPYRIGHT
       This script is licensed under the MIT License.


setup/ecocrumb.sh
#!/bin/bash

# Get the directory where the script is located
SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"

# Set the version
VERSION="v0.0.1"

# Check if the first argument is '--version'
if [ "$1" = "--version" ]; then
  echo "ecocrumb $VERSION"
  exit 0
fi

force=0
batch=0
train=0

# Parse command line options
while (( "$#" )); do
  case "$1" in
    --force)
      force=1
      shift
      ;;
    --batch)
      batch=1
      shift
      ;;
    --train)
      train=1
      shift
      ;;
    --) # end argument parsing
      shift
      break
      ;;
    -*|--*=) # unsupported flags
      echo "Error: Unsupported flag $1" >&2
      exit 1
      ;;
    *) # preserve positional arguments
      PARAMS="$PARAMS $1"
      shift
      ;;
  esac
done
eval set -- "$PARAMS"

# Check if the user provided the required argument
if [ $# -lt 1 ] || [ $# -gt 2 ]; then
  echo "Usage: ./ecocrumb.sh [--force] [--batch] [--train] <target_directory> [num_topics]"
  exit 1
fi

# Define the target directory or directories_path
target=$(realpath $1)

# Define the number of topics (default value: 15)
num_topics=${2:-15}

# If train mode is enabled, run the generate_topics.sh script
if [ "$train" -eq 1 ]; then
  "$SCRIPT_DIR/../esgetlpipeline/src/generate_topics.sh"
fi

# If batch mode is enabled, process each folder in the directory
if [ "$batch" -eq 1 ]; then
  for folder in "$target"/*; do
    if [ -d "$folder" ]; then
      echo "Processing folder: $folder"
      if [ "$force" -eq 1 ]; then
        "$SCRIPT_DIR/ecocrumb.sh" --force "$folder" "$num_topics"
      else
        "$SCRIPT_DIR/ecocrumb.sh" "$folder" "$num_topics"
      fi
      echo "Finished processing folder: $folder"
    fi
  done
  exit 0
fi

# Otherwise, process a single target directory
target_directory="$target"

# Define the output directories
topics_features_directory="${target_directory}/topics_features"
parsed_sections_directory="${target_directory}/parsed_sections"

# Create the output directories if they don't exist
mkdir -p "$topics_features_directory"
mkdir -p "$parsed_sections_directory"

# Define the output directory
output_directory=${target_directory}
if [ ! -d $output_directory ]; then
  mkdir $output_directory
fi

# Define the plaintext directory
plaintext_directory="${target_directory}/plaintext"
if [ ! -d $plaintext_directory ]; then
  mkdir $plaintext_directory
fi

# Ensure that the plaintext directory is not the same as the target directory
if [ "$plaintext_directory" = "$target_directory" ]; then
  echo "Error: The plaintext directory and the target directory are the same. Please check your directory structure."
  exit 1
fi

# Count the number of .txt files in the plaintext directory
txt_file_count=$(find "$plaintext_directory" -type f -name "*.txt" | wc -l)

# Convert all PDFs in the target directory to plaintext files only if there are no .txt files
if [ $txt_file_count -eq 0 ]; then
  echo "$(date +'%Y-%m-%d %H:%M:%S') - Converting all PDFs in $target_directory to plaintext files"
  echo "ECHO! - plaintext_directory: $plaintext_directory output_directory: $output_directory"
  # python main.py -i $target_directory -o $plaintext_directory
  $SCRIPT_DIR/.././esgetlpipeline/src/pdf_to_text.sh $target_directory $plaintext_directory
else
  echo "$(date +'%Y-%m-%d %H:%M:%S') - Skipping PDF conversion as .txt files already exist in $plaintext_directory"
fi

echo "Listing .txt files in $plaintext_directory:"
ls "$plaintext_directory"/*.txt
echo

# Process each plaintext file
for file in $plaintext_directory/*.txt; do
  echo "$(date +'%Y-%m-%d %H:%M:%S') - Processing file: $file"
  filename=$(basename $file .txt)

  topics_features_file="${topics_features_directory}/${filename}_topics_features.json"
  parsed_sections_file="${parsed_sections_directory}/${filename}_parsed_sections.json"

  echo "DEBUG: file variable: $file"
  echo "DEBUG: filename variable: $filename"
  echo "DEBUG: output_directory variable: $output_directory"
  echo "DEBUG: output JSON file: ${output_directory}/${filename}_topics_features.json"
  echo "$(date +'%Y-%m-%d %H:%M:%S') - Generating topics_features.json for $filename"

  # Check if the topics_features file already exists
  if [ ! -f "$topics_features_file" ] || [ "$force" -eq 1 ]; then
    # Generate topics_features.json
    echo "$(date +'%Y-%m-%d %H:%M:%S') - Generating topics_features.json for $filename"
    "$SCRIPT_DIR/../esgetlpipeline/src/run_topic_modeling.sh" "$file" $num_topics "$topics_features_file"
  else
    echo "$(date +'%Y-%m-%d %H:%M:%S') - Skipping topics_features generation for $filename, file already exists"
  fi

  # Check if the parsed_sections file already exists
  if [ ! -f "$parsed_sections_file" ] || [ "$force" -eq 1 ]; then
    # Generate parsed_sections.json
    echo "$(date +'%Y-%m-%d %H:%M:%S') - Generating parsed_sections.json for $filename"
    "$SCRIPT_DIR/../esgetlpipeline/src/parse_by_topic.sh" "$file" "$topics_features_file" "$parsed_sections_file"
  else
    echo "$(date +'%Y-%m-%d %H:%M:%S') - Skipping parsed_sections generation for $filename, file already exists"
  fi

done

# Copy contents of parsed_sections to results/ and esg_LanguageModel/src
cp -r "$parsed_sections_directory"/* "$SCRIPT_DIR/../results/"
cp -r "$parsed_sections_directory"/* "$SCRIPT_DIR/../esg_webapp/webapp/src/"
